core-java/src/main/java/org/spine3/server/EntityRepository.java

```diff
@@ -41,41 +45,53 @@
 
     @Nullable
     @Override
-    protected EntityStorage<I, M> getStorage() {
+    protected EntityStorage<I> getStorage() {
         @SuppressWarnings("unchecked") // It is safe to cast as we check the type in checkStorageClass().
-        final EntityStorage<I, M> storage = (EntityStorage<I, M>) super.getStorage();
+        final EntityStorage<I> storage = (EntityStorage<I>) super.getStorage();
         return storage;
     }
 
     @Override
-    public void store(E obj) {
-        final M state = obj.getState();
-        final EntityStorage<I, M> storage = checkStorage();
-        storage.write(obj.getId(), state);
+    public void store(E entity) {
+        final EntityStorage<I> storage = checkStorage();
+        final EntityRecord record = toEntityRecord(entity);
+        storage.store(record);
     }
 
     @Nullable
     @Override
     public E load(I id) {
-        final EntityStorage<I, M> storage = checkStorage();
-
-        final M entityState = storage.read(id);
-        if (entityState != null) {
-            final E result = create(id);
+        final EntityStorage<I> storage = checkStorage();
+        final EntityRecord record = storage.load(id);
+        if (record == null) {
+            return null;
+        }
+        final E entity = toEntity(id, record);
+        return entity;
+    }
 
-            //TODO:2015-09-28:alexander.yevsyukov: What do we pass for version and timestamp?
-            // Presumably EntityStorage should store these two fields and set it here.
+    @SuppressWarnings("TypeMayBeWeakened")
+    private E toEntity(I id, EntityRecord record) {
+        final E entity = create(id);
+        final M state = fromAny(record.getEntityState());
+        entity.setState(state, record.getVersion(), record.getWhenModified());
+        return entity;
+    }
 
-            result.setState(entityState, 0, TimeUtil.getCurrentTime());
-            return result;
-        }
-        // No state stored -> no entity.
-        return null;
+    private EntityRecord toEntityRecord(E entity) {
+        final String idString = idToString(entity.getId());
```

Comment:
Why do we have it as `string`? It can be one of the integer types too. We have a set of supported ID types, and we have the type we deal with defined as generic type parameter.